File: Java/Induttore.java
// classe concreta Induttore, derivata da Componente


// TO DO:
//   GESTIONE ECCEZIONI -> segnalate nell'implementazione dei vari metodi


//SCELTE PROGETTUALI:
// 1) override dei metodi toString(), equals(), impedenza()
// 3) "traduzione" in linguaggio Java delle ulteriori funzionalita' offerte dalla corrispettiva classe C++
//    a)costruttore da double
//    c)costruttore di default

class Induttore extends Componente{

  private double induttanza;

  public Induttore(double d){
    induttanza = d;
    //System.out.println(this.impedenza());
    super.setImp(impedenza());
  }

  public Induttore(){
    induttanza = 0;
  }

  public double getInduttanza(){
    return induttanza;
  }

  public CCartesiano impedenza(){
    double aux = 2 * (Numero.pi) * (Componente.freq) * (induttanza/1000);
    CCartesiano result = new CCartesiano(0, aux);
    return result;
  }

  /*
  public void assign(Dato d){
    if(!(d instanceof Induttore)){
      System.out.println("Induttore.assign()-> d non e' istanza di Induttore")
      //-------------------
      //GESTIRE ECCEZIONE
      //-------------------
    }

    Induttore aux = (Induttore) d;

    super.assign(aux);

    induttanza = aux.induttanza;
  }
  */

  public boolean equals(Object obj){
    if(!(obj instanceof Induttore)){
      return false;
    }

    Induttore aux = (Induttore) obj;

    if(super.equals(aux)){
      if(induttanza == aux.induttanza){
        return true;
      }
      else{
        return false;
      }
    }
    else{
      return false;
    }
  }

  public String toString(){
    String result = super.toString();
    result = result + "INDUTTANZA: " + induttanza + "\r\n";
    return result;
  }
}

File: KALK/Model/Induttore.cpp
#include "Induttore.h"

Induttore::Induttore(double i){
    if(i > 0)
        induttanza = i;
    else
        throw logic_exception("Inizializzato con valore negativo");
    try {
        Componente::setImp(impedenza());
    }
    catch(const syntax_exception& se){
        throw ;
    }
}

Induttore::Induttore(std::string s){
  auto pos=s.find('L');
  if(pos==0){
    try {
        induttanza = toDouble(s.substr(1));
        Componente::setImp(impedenza());
    }
    catch (const syntax_exception& se){
        throw;
    }
    catch (const std::exception& ex){
        throw syntax_exception("errore di sintassi");
    }
  }
  else
      throw syntax_exception("La stringa non inizia con il carattere L"); //gestire eccezione errore di sintassi
}

Induttore::Induttore():induttanza(0){}


bool Induttore::operator== (const Dato& d) const {
    try {
        auto aux = dynamic_cast<const Induttore &>(d);
        return induttanza == aux.induttanza;
    }
    catch (std::bad_cast &error){
        return false;
    }
}

std::string Induttore::toString() const{
    std::string res= "L"+std::to_string(getInduttanza());
    return res;
}

double Induttore::getInduttanza() const{
    return induttanza;
}

C_cartesiano Induttore::impedenza() const{
    double xl=2*Numero::pi*Componente::freq*(induttanza/1000);
    return C_cartesiano(0,xl);
}
